backends/p4tools/common/core/abstract_execution_state.h

```diff
@@ -87,11 +87,20 @@ class AbstractExecutionState {
     /* =========================================================================================
      *  General utilities involving ExecutionState.
      * ========================================================================================= */
+ protected:
+    /// Convert the input reference into a complex expression such as a HeaderExpression,
+    /// StructExpression, or HeaderStackExpression. @returns nullptr if the expression is not
+    /// complex.
+    const IR::Expression *convertToComplexExpression(const IR::StateVariable &parent) const;
+
+    /// Takes in a complex expression as a StructExpression, ListExpression, or
+    /// HeaderStackExpression, flattens it into a vector and @returns this vector. In parallel this
+    /// function fills a vector of the validity variables of any Type_Header encountered in this
+    /// expression.
+    static std::vector<const IR::Expression *> flattenComplexExpression(
+        const IR::Expression *inputExpression, std::vector<const IR::Expression *> &flatValids);
```

Comment:
I suggest returning a pair of vectors (or named struct) instead of using out-parameter. Out-params are often considered code smell in modern C++.

Reply:
Also would do this in a separate issue, fixing this for not just this function is a little more involved. 
#4266